### PR TITLE
[ML][Pipelines] Support expression in if-else

### PR DESCRIPTION
**Pull request by zhengfeiwang**
# Description

- resolve `PipelineExpression` in the if-else case
- fix bug by building expression component command following v2 contract

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_pipeline_expression.py [hunk 1]
```diff
@@ -542,9 +544,13 @@ def _generate_yaml_file(_path: Path) -> None:
...
-                _data["inputs"][_name] = {"type": _type}
-                _command_inputs_items.append(f"{_name}=\"$AZUREML_PARAMETER_{_name}\"")
+                _input = self._inputs[_name]
+                _data["inputs"][_name] = {"type": _input.type}
+                if isinstance(_input.value, NodeOutput):
+                    _param_name = f"{_input.value._owner.component.name}_{_input.value._name}"
+                else:
+                    _param_name = _name
+                _command_inputs_items.append(f"{_name}=\"$AZUREML_PARAMETER_{_param_name}\"")
             _command_inputs_string = " ".join(_command_inputs_items)
```
Comment (zhengfeiwang) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_pipeline_expression.py:554]: need to figure out how command generate
may still need `mldesigner` to generate component
breaking case: <https://ml.azure.com/runs/loyal_dolphin_9lgtblq41t?wsid=/subscriptions/96aede12-2f73-41cb-b983-6d11a904839b/resourcegroups/sdk/workspaces/sdk-master>

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_pipeline_expression.py [hunk 2]
```diff
@@ -334,17 +334,17 @@ def _handle_component_output(
                 _seen_input = _expression_inputs[_name]
                 if isinstance(_seen_input.value, PipelineInput):
                     if not _has_prefix:
-                        _name = f"{_component_output._owner.component.name}__{_component_output._name}"
+                        _name = f"{_component_output._owner.component.name}_{_component_output._name}"
```
Comment (elliotzh) [inline sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_pipeline_expression.py:337]: why this?

**Comment by elliotzh** [comment]
do we have any related tests?

**Comment by zhengfeiwang** [comment]
> do we have any related tests?

the related tests locate in mldesigner, I'll update them later

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py [hunk 3]
```diff
@@ -37,12 +38,15 @@ def pipeline_func(int_param1: int, int_param2: int):
...
+    :type condition: Union[str, bool, InputOutputBase, PipelineExpression]
     :param true_block: Block to be executed if condition resolved result is true.
     :type true_block: BaseNode
     :param false_block: Block to be executed if condition resolved result is false.
     :type false_block:  BaseNode
     :return: The condition node.
     :rtype: ConditionNode
     """
+    # resolve expression as command component
+    if isinstance(condition, PipelineExpression):
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py:50]: Seems this looks like the case: if user pass a node with only one output, we can do some similar logic to extract the single output

Comment (zhengfeiwang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py:50]: yes, is there any existing method to do so?

File: sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_pipeline_expression.py (hunk 2 above)
Comment (zhengfeiwang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_pipeline_expression.py:337]: offline synced on this, the root cause for this change is in some cases, the `command` for mldesigner command component is not expected for current code, need to align this and have some methods to sync this contract.

Comment (elliotzh) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_pipeline_expression.py]: offline synced

Comment (zhengfeiwang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/entities/_job/pipeline/_pipeline_expression.py]: I have checked related logic for mldesigner and find that I was wrong on the understanding of mldesigner component `command`: I used to think there is some special logic for mldesigner, but in fact there isn't and it is service-side execute related overwrite.
So here we can use something like `${{inputs.xxx}}/${{outputs.output}}` in `command` and everything will work fine under v2 contract.

**Comment by zhengfeiwang** [comment]
> do we have any related tests?

as current PR does not change component output parameter name, no need to update test here.
will add one for the resolve logic

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py (hunk 3 above)
Comment (wangchao1230) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py:50]: Can you add this logic here?

Comment (zhengfeiwang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py:52]: Refine this part as below:

1. convert `PipelineExpression` to a node
2. check the number of node's output, if not 1, raise `UserErrorException`
3. get the only output of the node

also add type `BaseNode` to param `condition`'s annotation

**Review by wangchao1230 — APPROVED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py [hunk 4]
```diff
@@ -6,12 +6,14 @@
...
 from azure.ai.ml.entities._job.pipeline._io import InputOutputBase
+from azure.ai.ml.entities._job.pipeline._pipeline_expression import PipelineExpression
+from azure.ai.ml.exceptions import UserErrorException
 
 # pylint: disable=redefined-outer-name
 
 
 def condition(
-    condition: Union[str, bool, InputOutputBase],
+    condition: Union[str, bool, InputOutputBase, BaseNode, PipelineExpression],
```
Comment (wangchao1230) [inline sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py:16]: no test in this PR?

Comment (zhengfeiwang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py:16]: good catch, let me add one for this

**Review by wangchao1230 — APPROVED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py [hunk 5]
```diff
@@ -36,13 +38,24 @@ def pipeline_func(int_param1: int, int_param2: int):
...
     :param false_block: Block to be executed if condition resolved result is false.
     :type false_block:  BaseNode
     :return: The condition node.
     :rtype: ConditionNode
     """
+    # resolve expression as command component
+    if isinstance(condition, PipelineExpression):
+        condition = condition.resolve()
+    if isinstance(condition, BaseNode):
+        if len(condition.outputs) != 1:
```
Comment (D-W-) [inline sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py:55]: Should we validate if the output is control output?

Comment (zhengfeiwang) [reply sdk/ml/azure-ai-ml/azure/ai/ml/dsl/_condition.py:55]: https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/ml/azure-ai-ml/azure/ai/ml/entities/_builders/condition_node.py#L61
seems we already have validation on this?

**Pull request merged**